azure-samples/src/main/java/com/microsoft/azure/management/compute/samples/CreateVirtualMachinesAsyncTrackingRelatedResources.java

```diff
@@ -0,0 +1,321 @@
+                            }
+                        }
+                        return createdResource;
+                    }
+               })
+               .onErrorReturn(new Func1<Throwable, Indexable>() {
+                   // Aggregate all the errors
+                   @Override
+                   public Indexable call(Throwable throwable) {
+                       errors.add(throwable);
+                       return null;
+                   }
+                })
+
+               // Start the process and wait for the Observable to stop emitting
+               .toBlocking().last();
+
+               System.out.println("Creation completed.");
+
+            // =====================================================================
+            // Clean up orphaned resources
+            //
+
+            // After everything has been created, first we delete the remaining successfully created NICs of failed VM creations
+            Collection<String> nicIdsToDelete = new ArrayList<>();
+            for (Creatable<NetworkInterface> nicDefinition : nicDefinitions.values()) {
+                String nicId = createdResourceIds.get(nicDefinition.key());
+                if (nicId != null) {
+                    nicIdsToDelete.add(nicId);
+                }
+            }
+            if (!nicIdsToDelete.isEmpty()) {
+                // Delete the NICs in parallel for better performance
+                azure.networkInterfaces().deleteByIds(nicIdsToDelete);
+            }
+
+            // Delete remaining successfully created resources of failed VM creations in parallel
+            Collection<Completable> deleteObservables = new ArrayList<>();
+            for (Collection<Creatable<? extends Resource>> relatedResources : vmNonNicResourceDefinitions.values()) {
+                for (Creatable<? extends Resource> resource : relatedResources) {
+                    String createdResourceId = createdResourceIds.get(resource.key());
+                    if (createdResourceId != null) {
+                        // Prepare the deletion of each related resource (treating it as a generic resource) as a multi-threaded Observable
+                        deleteObservables.add(azure.genericResources().deleteByIdAsync(createdResourceId).subscribeOn(Schedulers.io()));
+                    }
+                }
+            }
+
+            // Delete the related resources in parallel, as much as possible, postponing the errors till the end
+            Completable.mergeDelayError(deleteObservables).await();
+
+            // Show any errors
+            for (Throwable error : errors) {
+                System.out.println("\n### ERROR ###\n");
+                if (error instanceof CloudException) {
+                    CloudException ce = (CloudException) error;
+                    System.out.println("Cloud Exception: " + ce.getMessage());
+                } else {
+                    error.printStackTrace();
+                }
+            }
```

Comment:
Due to the reasons mentioned in the previous comment, we could use single error variable instead of list and replace the for loop with:

```java
if (error != null) {
    if (error instanceof rx.CompositeException) {
       for (Throwable t : ((rx.CompositeException)(error)).getExceptions()) {
           if (error instanceof CloudException) {
               CloudException ce = (CloudException) t;
                System.out.println("Cloud Exception: " + ce.getMessage())
           } else {
                t.printStackTrace();
           }
       }
   } else if (error instanceof CloudException) {
       CloudException ce = (CloudException) error;
       System.out.println("Cloud Exception: " + ce.getMessage())
   } else {
       error.printStackTrace();
   }
}
```

Reply:
Today, I understand that we bail out of the parallel creation as soon as we hit the first error. But in the past we talked about being more eager about that and continuing creating VMs, and just accumulate errors (`mergeDelayError()` semantics, rather than `merge()`, or something like that).  Wouldn't the current logic better protect against that potential future implementation?

Reply:
hmm, the current logic not really protect against outcome of potential future implementation where we continue creating resources eagerly. The reason is - in both cases, the we will be conforming to Observable rule where the terminal events `onError` or `onComplete` must invoked only once. Due to this rule, if there are multiple error happens (either due to parallel errors in case of `merge` or due to delaying of error due to `mergeDelayError`) internally they will be aggregated and thrown as Composite exception. So in nutshell, an observable will never throw more than once.